### PR TITLE
Wait key cert ready when auto mTLS is enabled and SDS is not enabled.

### LINKED ISSUE #18106: Gateway/Sidecar rejects the auto mTLS config when key cert is not ready
**Bug description**

TLDR, Pilot agent's readiness probe should mark the sidecar not ready util the first mTLS key cert is loaded from Citadel or other sources to avoid config push rejection.

This is more obvious w.r.t auto mTLS because key/cert is referenced in the CDS push right away...

**Solution**

Pilot agent its own readiness health check respect the key cert on its pod file system... Return failure if the key/cert is not ready.

we probably need to consider SDS as well somehow...

--- 

**Details**

When trying to turn on mTLS by default, many tests starts to fail. And interestingly they fall into some similar pattern.

- This test many involves some ingress/egress gateway.
- Adding some timeout retry will mitigate the test.

Let's take `make e2e_dashboard` as an example. The test set up some internal app `echosrv` and the go test program send request to `<INGRESS_IP>` and then with test provided virtual service route, send traffic to `echosrv`  app, expect request succeed or fails due to mixer policy check.

However, with mTLS enabled, we have to wait for about 30 seconds before the gateway becomes useful. Requests before that is always 503 NR (no route, per envoy doc).

So what causes the delay?

Look closely, we find in Pilot log

```log
2019-10-21T02:35:25.915472Z     warn    ads     ADS:CDS: ACK ERROR 10.56.13.92:44082 router~10.56.13.92~istio-ingressgateway-5959dfd8c-r5tjj.istio-system~istio-system.svc.cluster.local-2 (istio-ingressgateway-5959dfd8c-r5tjj.istio-system) Internal:Error adding/updating cluster(s) outbound|80||default-http-backend.kube-system.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem, outbound|80||heapster.kube-system.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem, outbound|53||kube-dns.kube-system.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem, outbound|443||metrics-server.kube-system.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem, outbound|10250||kubelet.kube-system.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem, outbound|443||kubernetes.default.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem, outbound|8000||httpbin.default.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem, outbound|80||istio-egressgateway.istio-system.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem, outbound|443||istio-egressgateway.istio-system.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem, outbound|15443||istio-egressgateway.istio-system.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem, outbound|443||istio-galley.istio-system.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem, outbound|15014||istio-galley.istio-system.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem, outbound|9901||istio-galley.istio-system.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem, outbound|15020||istio-ingressgateway.istio-system.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem, outbound|80||istio-ingressgateway.istio-system.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem, outbound|443||istio-ingressgateway.istio-system.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem, outbound|31400||istio-ingressgateway.istio-system.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem, outbound|15029||istio-ingressgateway.istio-system.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem, outbound|15030||istio-ingressgateway.istio-system.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem, outbound|15031||istio-ingressgateway.istio-system.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem, outbound|15032||istio-ingressgateway.istio-system.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem, outbound|15443||istio-ingressgateway.istio-system.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem, outbound|15010||istio-pilot.istio-system.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem, outbound|15011||istio-pilot.istio-system.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem, outbound|8080||istio-pilot.istio-system.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem, outbound|15014||istio-pilot.istio-system.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem, outbound|15004||istio-policy.istio-system.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem, outbound|15014||istio-policy.istio-system.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem, outbound|15004||istio-telemetry.istio-system.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem, outbound|15014||istio-telemetry.istio-system.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem, outbound|42422||istio-telemetry.istio-system.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem, outbound|9090||prometheus.istio-system.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem, outbound|8060||istio-citadel.istio-system.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem, outbound|15014||istio-citadel.istio-system.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem, outbound|443||istio-sidecar-injector.istio-system.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem, outbound|15014||istio-sidecar-injector.istio-system.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem, outbound|8080||echosrv.istio-system.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem, outbound|44444||netcat-srv.istio-system.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem, outbound|9999||netcat-client.istio-system.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem
2019-10-21T02:35:31.292950Z     info    ads     Push Status: {
    "ProxyStatus": {
        "pilot_eds_no_instances": {
            "outbound_.8000_._.httpbin.default.svc.cluster.local": {},
            "outbound|8000||httpbin.default.svc.cluster.local": {}
        }
    },
    "Version": "ck4II20H7bg="
}
2019-10-21T02:35:45.333501Z     info    ads     ADS:CDS: REQ 10.56.6.43:38090 router~10.56.6.43~istio-egressgateway-5cf588b56b-d8456.istio-system~istio-system.svc.cluster.local-6 715.637µs version:
2019-10-21T02:35:45.335487Z     info    ads     CDS: PUSH for node:istio-egressgateway-5cf588b56b-d8456.istio-system clusters:83 services:19 version:2019-10-21T02:35:25Z/17
2019-10-21T02:35:46.351374Z     info    ads     EDS: PUSH for node:istio-egressgateway-5cf588b56b-d8456.istio-system clusters:82 endpoints:126 empty:[outbound_.8000_._.httpbin.default.svc.cluster.local outbound|8000||httpbin.default.svc.cluster.local]
2019-10-21T02:35:46.447328Z     info    ads     LDS: PUSH for node:istio-egressgateway-5cf588b56b-d8456.istio-system listeners:0
2019-10-21T02:35:49.454685Z     info    transport: loopyWriter.run returning. connection error: desc = "transport is closing"
2019-10-21T02:35:49.454947Z     error   ads     ADS: "10.56.6.43:37964" router~10.56.6.43~istio-egressgateway-5cf588b56b-d8456.istio-system~istio-system.svc.cluster.local-1 terminated with stream closed
2019-10-21T02:35:49.456563Z     warn    ads     EDS: Send failure router~10.56.6.43~istio-egressgateway-5cf588b56b-d8456.istio-system~istio-system.svc.cluster.local-1: rpc error: code = Unavailable desc = transport is closing
2019-10-21T02:35:51.742167Z     info    ads     ADS:CDS: REQ 10.56.13.92:44300 router~10.56.13.92~istio-ingressgateway-5959dfd8c-r5tjj.istio-system~istio-system.svc.cluster.local-7 614.572µs version:
2019-10-21T02:35:51.743909Z     info    ads     CDS: PUSH for node:istio-ingressgateway-5959dfd8c-r5tjj.istio-system clusters:83 services:19 version:2019-10-21T02:35:25Z/17
2019-10-21T02:35:52.815877Z     info    ads     EDS: PUSH for node:istio-ingressgateway-5959dfd8c-r5tjj.istio-system clusters:82 endpoints:126 empty:[outbound_.8000_._.httpbin.default.svc.cluster.local outbound|8000||httpbin.default.svc.cluster.local]
2019-10-21T02:35:52.913664Z     info    ads     LDS: PUSH for node:istio-ingressgateway-5959dfd8c-r5tjj.istio-system listeners:1
2019-10-21T02:35:53.013309Z     info    ads     RDS: PUSH for node:istio-ingressgateway-5959dfd8c-r5tjj.istio-system routes:1
2019-10-21T02:35:56.114561Z     info    transport: loopyWriter.run returning. connection error: desc = "transport is closing"
2019-10-21T02:35:56.114619Z     error   ads     ADS: "10.56.13.92:44082" router~10.56.13.92~istio-ingressgateway-5959dfd8c-r5tjj.istio-system~istio-system.svc.cluster.local-2 terminated with stream closed
2019-10-21T02:35:56.114649Z     warn    ads     EDS: Send failure router~10.56.13.92~istio-ingressgateway-5959dfd8c-r5tjj.istio-system~istio-system.svc.cluster.local-2: rpc error: code = Unavailable desc = transport is closing

```

In gateway log
```
[2019-10-21 02:35:13.167][20][warning][config] [external/envoy/source/common/config/grpc_mux_subscription_impl.cc:82] gRPC config for type.googleapis.com/envoy.api.v2.Cluster rejected: Error adding/updating cluster(s) ou
tbound|80||default-http-backend.kube-system.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem, outbound|80||heapster.kube-system.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem, outbound|53||kube-dns.kube-system.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem, outbound|443||metrics-server.kube-system.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem, outbound|10250||kubelet.kube-system.svc.cluster.local: Invalid path: /etc/certs/cer
t-chain.pem, outbound|443||kubernetes.default.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem, outbound|8000||httpbin.default.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem, outbound|80||istio-egressgateway.istio-system.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem, outbound|443||istio-egressgateway.istio-system.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem, outbound|15443||istio-egressgateway.istio-system.svc.cluster.local
: Invalid path: /etc/certs/cert-chain.pem, outbound|443||istio-galley.istio-system.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem, outbound|15014||istio-galley.istio-system.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem, outbound|9901||istio-galley.istio-system.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem, outbound|15020||istio-ingressgateway.istio-system.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem, outbound|80||istio-i
ngressgateway.istio-system.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem, outbound|443||istio-ingressgateway.istio-system.svc.cluster.local: Invalid path: /etc/certs/cert-chain.pem,

2019-10-21T02:35:40.074719Z     info    watchFileEvents: "/etc/certs/..2019_10_21_02_35_40.074588345": CREATE                                                                                                                                 
2019-10-21T02:35:40.074899Z     info    watchFileEvents: "/etc/certs/..2019_10_21_02_35_40.074588345": MODIFY|ATTRIB                                                                                                                          
2019-10-21T02:35:40.074969Z     info    watchFileEvents: "/etc/certs/key.pem": CREATE                                                                                                                                                         
2019-10-21T02:35:40.074975Z     info    watchFileEvents: "/etc/certs/root-cert.pem": CREATE                                                                                                                                                   
2019-10-21T02:35:40.074980Z     info    watchFileEvents: "/etc/certs/cert-chain.pem": CREATE                                                                                                                                                  
2019-10-21T02:35:40.074996Z     info    watchFileEvents: "/etc/certs/..data_tmp": RENAME                                                                                                                                                      
2019-10-21T02:35:40.075001Z     info    watchFileEvents: "/etc/certs/..data": CREATE                                                                                                                                                          
2019-10-21T02:35:40.075016Z     info    watchFileEvents: "/etc/certs/..2019_10_21_02_34_29.329176922": DELETE                                                                                                                                 
[2019-10-21T02:35:44.292Z] "GET / HTTP/1.1" 503 NR "-" "-" 0 0 4 - "10.128.0.48" "Go-http-client/1.1" "0a3b74b9-dcf2-468d-ba80-6a8db358d4af" "35.202.167.158" "-" - - 10.56.13.92:80 10.128.0.48:46282 - -                                    
2019-10-21T02:35:50.074985Z     info    watchFileEvents: notifying                                                                                                                                                                            
2019-10-21T02:35:50.075180Z     info    Received new config, creating new Envoy epoch 1                                                                                                                                                       
2019-10-21T02:35:50.075233Z     info    waiting for epoch 0 to go live before performing a hot restart                                                                                                                                        
2019-10-21T02:35:50.076077Z     info    Epoch 1 starting


```

So the auto mTLS CDS config is delivered as expected, but the gateway Envoy can't accept it because the referenced key/cert is not present in Envoy file system (Good job Envoy's sanity check :)

After Citadel deliver the file mount to the gateway, we find the config push is accepted, request starts to succeed as before... in a sequence.

**Affected product area (please put an X in all that apply)**

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[X ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[X ] Security
[ X] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Expected behavior**

**Steps to reproduce the bug**

**Version (include the output of `istioctl version --remote` and `kubectl version`)**

**How was Istio installed?**

**Environment where bug was observed (cloud vendor, OS, etc)**

Additionally, please consider attaching a [cluster state archive](http://istio.io/help/bugs/#generating-a-cluster-state-archive) by attaching
the dump file to this issue.

### PR DESCRIPTION
**Pull request by incfly**
Please provide a description for what this PR is for.

And to help us figure out who should review this PR, please 
put an X in all the areas that this PR affects.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Fixed https://github.com/istio/istio/issues/18106

### CODE REVIEW

**Comment by incfly** [comment]
@howardjohn @linsun could you take a look of this PR? We need this to be merged in order to unblock https://github.com/istio/istio/pull/18030/files

I think it's better to have separate PR from above, because know we introduce implementation change in pilot agent.

Feel free to appoint other reviewer candidates.

File: pilot/cmd/pilot-agent/status/ready/probe.go [hunk 1]
```diff
@@ -48,9 +59,22 @@ func (p *Probe) Check() error {
 		}
 	}
 
+	// Check key cert on file system as `/etc/cert/key.pem`
```
Comment (howardjohn) [inline pilot/cmd/pilot-agent/status/ready/probe.go]: I think this may be configurable?

**Review by howardjohn — COMMENTED** [review]
Generally looks good, would like @costinm to look at this with regards to SDS stuff

File: pilot/cmd/pilot-agent/status/ready/probe.go (hunk 1 above)
Comment (incfly) [reply pilot/cmd/pilot-agent/status/ready/probe.go]: removed.

**Comment by incfly** [comment]
Changed to use existing key/cert  wait logic when auto mTLS is enabled.

File: pilot/cmd/pilot-agent/main.go [hunk 2]
```diff
@@ -99,12 +99,14 @@ var (
...
-	istioNamespaceVar         = env.RegisterStringVar("ISTIO_NAMESPACE", "", "")
-	kubeAppProberNameVar      = env.RegisterStringVar(status.KubeAppProberEnvName, "", "")
-	sdsEnabledVar             = env.RegisterBoolVar("SDS_ENABLED", false, "")
+	instanceIPVar        = env.RegisterStringVar("INSTANCE_IP", "", "")
+	podNameVar           = env.RegisterStringVar("POD_NAME", "", "")
+	podNamespaceVar      = env.RegisterStringVar("POD_NAMESPACE", "", "")
+	istioNamespaceVar    = env.RegisterStringVar("ISTIO_NAMESPACE", "", "")
+	kubeAppProberNameVar = env.RegisterStringVar(status.KubeAppProberEnvName, "", "")
+	sdsEnabledVar        = env.RegisterBoolVar("SDS_ENABLED", false, "")
+	autoMTLSEnabled      = env.RegisterBoolVar("ISTIO_AUTO_MTLS_ENABLED", false, "If true, auto mTLS is enabled, "+
```
Comment (howardjohn) [inline pilot/cmd/pilot-agent/main.go]: Why just auto mtls and not regular mtls?

Comment (incfly) [reply pilot/cmd/pilot-agent/main.go]: At least for now, regular mTLS is configured via authn policy, CRD, dynamically. We don't know whether a particular workload needs mTLS or not.

Having that said, we probably can trim lots of logic in future, if we can prove auto mTLS stability and even the authn policy for mTLS can be nuked. That'll be feature improvements.

**Review by howardjohn — APPROVED** [review]

**Pull request merged**